hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/PrngSystemContract.java

```diff
@@ -0,0 +1,145 @@
+
+        // compute the gas requirement
+        gasRequirement =
+                calculateGas(Instant.ofEpochSecond(frame.getBlockValues().getTimestamp()));
+
+        // get the contract ID
+        final ContractID contractID = asEvmContractId(Address.fromHexString(PRNG_PRECOMPILE_ADDRESS));
+
+        try {
+            // compute the pseudorandom number
+            final var randomNum = generatePseudoRandomData(input, frame);
+            requireNonNull(randomNum);
+            final var result = PrecompiledContract.PrecompileContractResult.success(randomNum);
+
+            // create a child record
+            createSuccessfulRecord(frame, randomNum, contractID);
+
+            return result;
+        } catch (Exception e) {
+            log.warn("Internal precompile failure", e);
+            // create a failed record and returned a halt result
+            createFailedRecord(frame, FAIL_INVALID.toString(), contractID);
+            return PrecompiledContract.PrecompileContractResult.halt(Bytes.EMPTY, Optional.of(INVALID_OPERATION));
+        }
+    }
+
+    void createSuccessfulRecord(
+            @NonNull MessageFrame frame, @NonNull final Bytes randomNum, @NonNull final ContractID contractID) {
+        if (!frame.isStatic()) {
+            requireNonNull(frame);
+            requireNonNull(randomNum);
+            requireNonNull(contractID);
+            var updater = (ProxyWorldUpdater) frame.getWorldUpdater();
+            updater.externalizeSystemContractResults(
+                    contractFunctionResultSuccessFor(gasRequirement, randomNum, contractID), ResultStatus.IS_SUCCESS);
+        }
+    }
+
+    void createFailedRecord(
+            @NonNull MessageFrame frame, @NonNull final String errorMsg, @NonNull final ContractID contractID) {
+        if (!frame.isStatic()) {
+            requireNonNull(frame);
+            requireNonNull(contractID);
+            contractFunctionResultFailedFor(gasRequirement, errorMsg, contractID);
+            var updater = (ProxyWorldUpdater) frame.getWorldUpdater();
+            updater.externalizeSystemContractResults(
+                    contractFunctionResultFailedFor(gasRequirement, errorMsg, contractID), ResultStatus.IS_ERROR);
+        }
+    }
+
+    Bytes generatePseudoRandomData(@NonNull final Bytes input, @NonNull final MessageFrame frame) {
+        final var selector = input.getInt(0);
+        if (selector == PSEUDORANDOM_SEED_GENERATOR_SELECTOR) {
+            return random256BitGenerator(frame);
+        }
+        return null;
+    }
+
+    Bytes random256BitGenerator(final MessageFrame frame) {
+        final var entropy = ((ProxyWorldUpdater) frame.getWorldUpdater()).entropy();
+        if (entropy.equals(Bytes.wrap(ZERO_ENTROPY.toByteArray()))) {
+            return null;
+        }
+        return entropy;
+    }
+
+    long calculateGas(@NonNull final Instant now) {
+        // TODO: Update gas calculations once the fee calculator classes are available
```

Comment:
May want to create a ticket for this and annotate: `@future('####')` (with the ticket number replacing '####')